### PR TITLE
add documentation

### PR DESCRIPTION
**Pull request by artemeff**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,24 +1,278 @@
...
+  @behaviour Struct.Type
+
+  @spec cast(term) :: {:ok, term} | {:error, term} | :error
+  def cast(value) do
+    {:ok, value}
+  end
+end
+```
+
+Notice that `cast/1` can return error with reason, that behaviour supports only in struct and you can't use this types in Ecto schemas.
```
Comment (spscream) [inline README.md]: this behaviour is supported only by Struct and you can't use types defined using Struct in Ecto schemas

File: README.md [hunk 2]
```diff
@@ -1,24 +1,278 @@
...
+  end
+end
+```
+
+```elixir
+iex> Quiz.make(%{user_id: 42, answers: ["yes", "no", "no", "yes"]})
+{:ok, %Quiz{answers: [true, false, false, true], user_id: 42}}
+```
+
+What if we need to parse 'optimized' query string from URL? Like list of user_id separated by a comma? Create a custom type for each boxed type? No! Use type composition:
```
Comment (spscream) [inline README.md]: What if we need to parse 'optimized' query string from URL, like list of user ids separated by a comma?
Do we need to create a custom type for each boxed type? No! Just use type composition feature:

File: README.md [hunk 3]
```diff
@@ -1,24 +1,278 @@
...
+{:error, %{comments: :missing, title: :missing}}
+
+iex> Post.make(%{comments: %{}, title: :test})
+{:error, %{comments: :invalid, title: :invalid}}
+
+iex> Post.make(%{comments: [%{}], title: "what the title?"})
+{:error, %{comments: %{text: :missing}}}
+```
+
+Or receive an exception with ivalid data:
```
Comment (spscream) [inline README.md]: with invalid data

File: lib/struct.ex [hunk 4]
```diff
@@ -61,7 +103,30 @@ defmodule Struct do
...
+  For custom types checks for module existence and `c:Struct.Type.cast/1` callback.
+
+  If field definition is invalid for some reason — it throws an `Struct.DefinitionError`
+  exception with detailed reason.
+
+  ## Options
+
+    * `:default` — sets default value for that field:
+
+      * Provided value from params compares with default value before and after type cast;
```
Comment (spscream) [inline lib/struct.ex]: Value from params is compared with default value before and after type cast;

File: lib/struct.ex [hunk 5]
```diff
@@ -61,7 +103,30 @@ defmodule Struct do
...
+
+      * If you pass `field :a, type, default: nil` and `make(%{a: nil})` — type coercion not even used,
```
Comment (spscream) [inline lib/struct.ex]: type coercion will not be used

File: lib/struct.ex [hunk 6]
```diff
@@ -61,7 +103,30 @@ defmodule Struct do
...
+        `nil` compares with default value and just appends that value to struct;
+
+      * If that field doesn't exist in params (missing value) — it just appends default value.
```
Comment (spscream) [inline lib/struct.ex]: If field doesn't exist in params, it will use default value

File: lib/struct.ex [hunk 7]
```diff
@@ -61,7 +103,30 @@ defmodule Struct do
...
+
+      By default this option is unset. Notice that you can't use functions as default value.
```
Comment (spscream) [inline lib/struct.ex]: as a default value

File: lib/struct.ex [hunk 8]
```diff
@@ -74,6 +139,23 @@ defmodule Struct do
     end
   end
 
+  @doc """
+  Alias to `Struct.Cast.make/3`.
+  """
+  @callback make(params :: map, opts :: Keyword.t) :: {:ok, t} | {:error, term}
+
+  @doc """
+  Alias to `c:make/2`, but raise an `Struct.MakeError` exception if params has errors.
```
Comment (spscream) [inline lib/struct.ex]: but raises a `Struct.MakeError` exception if params have errors.

File: lib/struct/cast.ex [hunk 9]
```diff
@@ -1,7 +1,60 @@
 defmodule Struct.Cast do
+  @moduledoc """
+  Module to make struct instance from provided params.
+
+  You can event use it standalone without defining struct, by providing types and params to `make/3`.
```
Comment (spscream) [inline lib/struct/cast.ex]: Что ты тут хотел сказать, напиши по русски плз

Comment (spscream) [reply lib/struct/cast.ex]: You can use it standalone, without defining struct, by providing types and params to `make/3` наверное

**Review by spscream — CHANGES_REQUESTED** [review]

File: lib/struct/cast.ex (hunk 9 above)
Comment (artemeff) [reply lib/struct/cast.ex]: event -> even, по привычке лишнее печатается

**Pull request merged**